Auto-generate the Cloud IoT Core Device Registry resource

ferrarimarco opened the pull request:
This PR automatically generates the [Cloud IoT Core](https://cloud.google.com/iot-core) [Device Registry](https://cloud.google.com/iot/docs/reference/cloudiot/rest/v1/projects.locations.registries) terraform resource, and the relevant acceptance tests. The acceptance tests cover every aspect that was considered by the previous test suite, plus a couple of additional scenarios.

The resource is currently manually generated ([src](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/third_party/terraform/resources/resource_cloudiot_registry.go), [test](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/third_party/terraform/tests/resource_cloudiot_registry_test.go), [markdown docs](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/third_party/terraform/website/docs/r/cloudiot_registry.html.markdown)). This PR takes care of removing these files, and sets the `legacy_name` field for the newly added product and resource to be backward compatible with the current implementation.

This is the first step in order to support https://github.com/terraform-providers/terraform-provider-google/issues/1495.

<!-- AUTOCHANGELOG for Downstream PRs.

EXTERNAL CONTRIBUTORS: Ignore please - your reviewer will handle.

INTERNAL CONTRIBUTORS AND REVIEWERS: See .ci/RELEASE_NOTES_GUIDE.md
for writing good release notes.

NO CHANGELOG NOTE: Please add "changelog: no-release-note" label to this PR.

Otherwise, fill the template out (replace the heading).
You can add more release notes if you want more than one CHANGELOG entry for
this PR, but make sure not to indent notes and to leave newlines between
code blocks for Markdown's sake.

For Terraform PRs, we use the following "release-note:" headings
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
-->

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 11 files changed, 1864 insertions(+), 858 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 11 files changed, 1864 insertions(+), 858 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 2 files changed, 371 insertions(+))

modular-magician commented:
Oops! It looks like you're using an unknown release-note type in your changelog entries:
* REPLACEME

Please only use the types listed in https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md.

ferrarimarco commented:
I don't have access to that Cloud Build job, so I cannot see why it's failing.

ferrarimarco commented:
Thanks @danawillow ! Will have a look at your comments.

I actually need to generate the Device resource, and took the chance to generate the registry as well, because I had the manually generated one as a reference to guide me,  while learning how MM work.

ferrarimarco commented:
Question: from what I see, this auto-generate resource might also fix the #6085 you just created, right?

Fixing #6085 requires to introduce a breaking change, so we might just use this PR to piggyback that fix. Or not?

ferrarimarco commented:
Or should I make this backward compatible as you suggest, but it has the same issue, and then the fix (in this case meaning the removal of the excludes and additional schema elements) as part of the 4.0.0 release?

danawillow commented:
This PR as it is now would fix that issue, but we can't make the breaking change until the 4.0.0 release. We don't expect to do that until after we've migrated to our new generator, so at that point this PR would be obsolete.

ferrarimarco commented on third_party/terraform/utils/provider.go.erb:
```diff
@@ -289,7 +289,6 @@ end     # products.each do
 				"google_billing_account_iam_member":            ResourceIamMember(IamBillingAccountSchema, NewBillingAccountIamUpdater, BillingAccountIdParseFunc),
 				"google_billing_account_iam_policy":            ResourceIamPolicy(IamBillingAccountSchema, NewBillingAccountIamUpdater, BillingAccountIdParseFunc),
 				"google_cloudfunctions_function":               resourceCloudFunctionsFunction(),
-				"google_cloudiot_registry":                     resourceCloudIoTRegistry(),
```
done

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 12 files changed, 1864 insertions(+), 871 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 12 files changed, 1864 insertions(+), 871 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 371 insertions(+), 11 deletions(-))

modular-magician commented:
Oops! It looks like you're using an unknown release-note type in your changelog entries:
* REPLACEME

Please only use the types listed in https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 12 files changed, 1286 insertions(+), 871 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 12 files changed, 1286 insertions(+), 871 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 147 insertions(+), 11 deletions(-))

modular-magician commented:
Oops! It looks like you're using an unknown release-note type in your changelog entries:
* REPLACEME

Please only use the types listed in https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md.

ferrarimarco commented on third_party/terraform/tests/resource_cloudiot_device_registry_id_test.go:
```diff
@@ -0,0 +1,30 @@
+package google
```
@danawillow Is `third_party/terraform/tests` the right place for additional tests? I don't see the included in the provider I generate.

ferrarimarco replied:
NVM, found them

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 12 files changed, 1333 insertions(+), 873 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 12 files changed, 1333 insertions(+), 873 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 147 insertions(+), 12 deletions(-))

modular-magician commented:
Oops! It looks like you're using an unknown release-note type in your changelog entries:
* REPLACEME

Please only use the types listed in https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 12 files changed, 1345 insertions(+), 873 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 12 files changed, 1345 insertions(+), 873 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 137 insertions(+), 12 deletions(-))

modular-magician commented:
Oops! It looks like you're using an unknown release-note type in your changelog entries:
* REPLACEME

Please only use the types listed in https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md.

ferrarimarco commented on products/cloudiot/api.yaml:
```diff
@@ -0,0 +1,194 @@
+              required: true
+              description: |
+                PubSub topic name to publish device events.
+      - !ruby/object:Api::Type::NestedObject
```
done

ferrarimarco commented on products/cloudiot/api.yaml:
```diff
@@ -0,0 +1,194 @@
+            required: true
+            description: |
+              PubSub topic name to publish device state updates.
+      - !ruby/object:Api::Type::NestedObject
```
done

ferrarimarco commented on products/cloudiot/api.yaml:
```diff
@@ -0,0 +1,194 @@
+            values:
+              - :MQTT_ENABLED
+              - :MQTT_DISABLED
+      - !ruby/object:Api::Type::NestedObject
```
done

ferrarimarco commented on products/cloudiot/api.yaml:
```diff
@@ -0,0 +1,194 @@
+        max_size: 10
+        item_type: !ruby/object:Api::Type::NestedObject
+          properties:
+            - !ruby/object:Api::Type::NestedObject
```
done

ferrarimarco commented:
@danawillow I've addressed your comments, thanks!

After running the acceptance tests on my side, I see that some of them are failing due to miscalculated diffs, even though I added a `diff_suppress_func` in the `terraform.yaml`.

Can you please have a look? Thanks!

danawillow commented on products/cloudiot/terraform.yaml:
```diff
@@ -0,0 +1,141 @@
+        # Excluding this because the original (manually-generated) implementation
+        # wrongly set this to be a Map, instead of a NestedObject. To avoid breaking
+        # changes, we observe that behaviour by excluding this field and adding a
+        # corresponding custom element (of type Map) to the schema. When we're
```
It's not just a custom schema entry that you have to add, but also the Create, Read, and Update logic for that entry in encoders/decoders. If you take a look at the generated code, you can see that it doesn't actually include any logic around the new fields; it just puts them in the schema.

Just as a reminder, in case the amount of work is getting to be too much: you don't have to do this. It's totally fine to leave the resource as handwritten and add new features to it there; we do this all the time.

ferrarimarco replied:
Right! Ok, will have a look at this. Do you know if there are other resources that are doing a similar thing, that I can use as an example? Thanks!

danawillow replied:
You can take a look at the encoders/ and decoders/ folder to see, but I'm not sure any of them will be as extensive as you'd need for this. Normally we would just leave this resource handwritten.

ferrarimarco replied:
done

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1437 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1437 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 188 insertions(+), 12 deletions(-))

ferrarimarco commented:
Hi @danawillow! I've added custom encoders and decoders for `http_config`, `mqtt_config`, `state_notification_config`.

I still need to implement the ones for `credentials`, but I wanted to check if I'm going in the right direction.

I still see issues in the acceptance tests (tried the simplest one: `TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample`). Can you have a look?

Thanks!

```shell
make testacc TEST=./google TESTARGS='-run=TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample
=== PAUSE TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample
=== CONT  TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample
    TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample: testing.go:635: Step 0 error: After applying this step, the plan was not empty:

        DIFF:

        UPDATE: google_cloudiot_registry.test-registry
          credentials.#:                               "0" => "0"
          event_notification_configs.#:                "0" => "0"
          http_config.http_enabled_state:              "HTTP_ENABLED" => "HTTP_ENABLED"
          id:                                          "projects/magic-modules-iot-core-test/locations/us-central1/registries/tf-test-cloudiot-registryocownfcqz7" => "projects/magic-modules-iot-core-test/locations/us-central1/registries/tf-test-cloudiot-registryocownfcqz7"
          log_level:                                   "" => ""
          mqtt_config.mqtt_enabled_state:              "MQTT_ENABLED" => "MQTT_ENABLED"
          name:                                        "tf-test-cloudiot-registryocownfcqz7" => "tf-test-cloudiot-registryocownfcqz7"
          project:                                     "magic-modules-iot-core-test" => "magic-modules-iot-core-test"
          region:                                      "us-central1" => "us-central1"
          state_notification_config.pubsub_topic_name: "" => ""



        STATE:

        google_cloudiot_registry.test-registry:
          ID = projects/magic-modules-iot-core-test/locations/us-central1/registries/tf-test-cloudiot-registryocownfcqz7
          provider = provider.google
          http_config.http_enabled_state = HTTP_ENABLED
          log_level =
          mqtt_config.mqtt_enabled_state = MQTT_ENABLED
          name = tf-test-cloudiot-registryocownfcqz7
          project = magic-modules-iot-core-test
          region = us-central1
          state_notification_config.pubsub_topic_name =

          Dependencies:
            google_project_service.cloud-iot-apis
        google_project_service.cloud-iot-apis:
          ID = magic-modules-iot-core-test/cloudiot.googleapis.com
          provider = provider.google
          disable_dependent_services = true
          disable_on_destroy = true
          project = magic-modules-iot-core-test
          service = cloudiot.googleapis.com
--- FAIL: TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample (52.17s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-google/google	53.012s
FAIL
make: *** [testacc] Error 1
```

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1568 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1568 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 278 insertions(+), 12 deletions(-))

danawillow commented on products/cloudiot/terraform.yaml:
```diff
@@ -0,0 +1,144 @@
+      mqttConfig: !ruby/object:Overrides::Terraform::PropertyOverride
+        # See the comment on stateNotificationConfig.exclude
+        exclude: true
+        name: 'mqtt_config'
```
Why are you changing the name on fields you're excluding? Also I don't think this line will do anything since we'll case it correctly in the generator.

ferrarimarco replied:
Ok!

ferrarimarco replied:
done

danawillow commented on templates/terraform/examples/cloudiot_device_registry_basic.tf.erb:
```diff
@@ -0,0 +1,13 @@
+resource "google_project_service" "cloud-iot-apis" {
```
Could this be removed? These run as tests in our CI and I don't like the idea that the service is going  to get enabled and disabled during the test (plus it's inconsistent with our other resources)

ferrarimarco replied:
done, also in other tests

danawillow commented on templates/terraform/update_encoder/cloudiot_device_registry.go.erb:
```diff
@@ -0,0 +1,27 @@
+config := meta.(*Config)
```
Didn't look too closely- is this the exact same code as the create encoder? If so, I'd recommend using the same file for both in order to have fewer things that need to be maintained in the future.

ferrarimarco replied:
done

danawillow commented on third_party/terraform/tests/resource_cloudiot_device_registry_update_test.go:
```diff
@@ -0,0 +1,162 @@
+func TestAccCloudIoTRegistry_update(t *testing.T) {
+	t.Parallel()
+
+	registryName := fmt.Sprintf("psregistry-test-%s", acctest.RandString(10))
```
```suggestion
	registryName := fmt.Sprintf("psregistry-test-%s", randString(t, 10))
```

ferrarimarco replied:
done

danawillow commented on third_party/terraform/tests/resource_cloudiot_device_registry_update_test.go:
```diff
@@ -0,0 +1,162 @@
+
+	registryName := fmt.Sprintf("psregistry-test-%s", acctest.RandString(10))
+
+	resource.Test(t, resource.TestCase{
```
```suggestion
	vcrTest(t, resource.TestCase{
```

ferrarimarco replied:
done

danawillow commented on third_party/terraform/tests/resource_cloudiot_device_registry_update_test.go:
```diff
@@ -0,0 +1,162 @@
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckCloudIoTRegistryDestroy,
```
```suggestion
		CheckDestroy: testAccCheckCloudIotDeviceRegistryDestroyProducer(t),
```

ferrarimarco replied:
done

danawillow commented on third_party/terraform/tests/resource_cloudiot_device_registry_update_test.go:
```diff
@@ -0,0 +1,162 @@
+	})
+}
+
+func testAccCheckCloudIoTRegistryDestroy(s *terraform.State) error {
```
You shouldn't need this since you can just use the one from the generated test

ferrarimarco replied:
Nice! I didn't know about this. Will have a closer look at the generated provider

ferrarimarco replied:
done

danawillow commented on third_party/terraform/tests/resource_cloudiot_device_registry_update_test.go:
```diff
@@ -0,0 +1,162 @@
+    }
+  }
+}
+`, acctest.RandString(10), acctest.RandString(10), acctest.RandString(10), registryName, registryName)
```
can you create a new random variable in the test using `randString(t, 10)` and then pass it here as a parameter instead? we're doing some test overhauls and that's one of the things the  new framework needs

ferrarimarco replied:
done

danawillow commented on products/cloudiot/api.yaml:
```diff
@@ -0,0 +1,194 @@
+            values:
+              - :HTTP_ENABLED
+              - :HTTP_DISABLED
+      - !ruby/object:Api::Type::Enum
```
Can you add `send_empty_value: true` to this field to line up with https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_cloudiot_registry.go#L250?

ferrarimarco replied:
done

ferrarimarco replied:
If I add `send_empty_value: true` to `logLevel`, my `TestAccCloudIoTRegistry_update` breaks, because it sends an empty `log_level` in the HTTP request, and the API doesn't like that (that field is an Enum that doesn't accept empty strings).

Should I remove that field from the updateMask if it's set to an empty string (or if the user doesn't set it)? Or am I missing something?

thanks!

```
---[ REQUEST ]---------------------------------------
PATCH /v1/projects/xxxxxxxxxxxxx/locations/us-central1/registries/psregistry-test-4d3gub0mql?alt=json&updateMask=eventNotificationConfigs%2Cstate_notification_config.pubsub_topic_name%2Cmqtt_config.mqtt_enabled_state%2Chttp_config.http_enabled_state%2Ccredentials HTTP/1.1
Host: cloudiot.googleapis.com
User-Agent: HashiCorp Terraform/0.12.7-sdk (+https://www.terraform.io) Terraform Plugin SDK/1.4.0 terraform-provider-google/dev
Content-Length: 1704
Content-Type: application/json
Accept-Encoding: gzip

{
 "credentials": [
  {
   "publicKeyCertificate": {
    "certificate": "-----BEGIN CERTIFICATE-----\naAAAAAAAAAA\n-----END CERTIFICATE-----\n",
    "format": "X509_CERTIFICATE_PEM"
   }
  }
 ],
 "eventNotificationConfigs": [
  {
   "pubsubTopicName": "projects/xxxxxxxxxxxxxxxxx/topics/psregistry-additional-test-telemetry-psregistry-additional-test-telemetry-iinb3u43lq",
   "subfolderMatches": "test/directory"
  },
  {
   "pubsubTopicName": "projects/xxxxxxxxxxxxxxx/topics/psregistry-test-telemetry-psregistry-test-telemetry-2o11c0orz4"
  }
 ],
 "httpConfig": {
  "httpEnabledState": "HTTP_DISABLED"
 },
 "logLevel": "",
 "mqttConfig": {
  "mqttEnabledState": "MQTT_DISABLED"
 },
 "stateNotificationConfig": {
  "pubsubTopicName": "projects/xxxxxxxxx/topics/psregistry-test-devicestatus-psregistry-test-devicestatus-kxqivokfwo"
 }
}
```

```
---[ RESPONSE ]--------------------------------------
HTTP/1.1 400 Bad Request
Connection: close
Transfer-Encoding: chunked
Alt-Svc: xxxxxx
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Thu, 07 May 2020 22:30:34 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

1cb
{
  "error": {
    "code": 400,
    "message": "Invalid value at 'device_registry.log_level' (TYPE_ENUM), \"\"",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.BadRequest",
        "fieldViolations": [
          {
            "field": "device_registry.log_level",
            "description": "Invalid value at 'device_registry.log_level' (TYPE_ENUM), \"\""
          }
        ]
      }
    ]
  }
}
```

danawillow replied:
Here's what I think is going on! The handwritten version of this resource has a few typos in it that nobody ever caught:
* https://github.com/terraform-providers/terraform-provider-google/blob/93047df43006806540f4fe6c013ccf5825be370e/google/resource_cloudiot_registry.go#L250 should be LogLevel
* https://github.com/terraform-providers/terraform-provider-google/blob/93047df43006806540f4fe6c013ccf5825be370e/google/resource_cloudiot_registry.go#L333 should be logLevel

So it looked like we were sending the empty value, but we actually weren't. It also looked like we were sending a valid updateMask, but we weren't. My guess is that if the update mask isn't valid, the API ignores it and treats it like a PUT (I'm not quite sure why, but that does seem to be what's happening).

This means that sending an empty log level would have accidentally worked (because the API doesn't send a log level back if it wasn't sent in the request), and updating a log level to empty would also have accidentally worked (because the API was replacing the whole resource with what we sent).

What I think we should actually do is set a default value of "NONE", which I believe from looking at the docs is what the API does. Then I think we can remove the DiffSuppressFunc, but we may need to replace it with `emptyOrDefaultStringSuppress("NONE")`. I'll let you figure that out. And then don't set `send_empty_value`.

This also made me realize that some of the updateMask values in your handwritten code were written as snake_case, but should be camelCase.

Give that all a try and let me know how it goes!

ferrarimarco replied:
Thanks for figuring this out! I'm working on this change now. Happy to see that we're going to fix a bug as well!

ferrarimarco replied:
@danawillow I've pushed a couple of commits with the requested changes:

1. Set the default value for log level to `:NONE` in `api.yaml`.
1. Don't set `send_empty_value` for log level.
1. Change `emptyOrDefaultStringSuppress("")` to `emptyOrDefaultStringSuppress("NONE")`.
1. Fix case of some values when populating the update mask.

All my test cases pass!

ferrarimarco replied:
(and Cloud Build is happy too!)

danawillow commented on products/cloudiot/terraform.yaml:
```diff
@@ -0,0 +1,144 @@
+    import_format: ["{{project}}/locations/{{region}}/registries/{{name}}"]
+    id_format: "projects/{{project}}/locations/{{region}}/registries/{{name}}"
+    self_link: 'projects/{{project}}/locations/{{region}}/registries/{{name}}'
+    custom_code: !ruby/object:Provider::Terraform::CustomCode
```
don't you also need a pre_update here in order to set the updateMask for the custom fields?

ferrarimarco replied:
done

danawillow commented:
Cloud build failure:
```
==> Checking source code against linters...
google/resource_cloud_iot_device_registry.go:79:6: `expandCloudIotDeviceRegistryStateNotificationConfigPubsubTopicName` is unused (deadcode)
func expandCloudIotDeviceRegistryStateNotificationConfigPubsubTopicName(v interface{}, d TerraformResourceData, config *Config) (interface{}, error) {
     ^
make: *** [lint] Error 1
```

I'm not sure about the acceptance test failure; the two sides of that update look identical to me. Once you make the changes to not enable/disable APIs during the tests, I can run them myself and take a closer look. I'll be doing that by setting TF_LOG=DEBUG and just looking at the API requests/responses and anything else interesting in the logs, so feel free to get a head start and do the same!  As another reminder, this entire change is just for aesthetics and is not necessary in 
 order to complete any other IoT resources, so I don't want to spend too much more time reviewing it.

ferrarimarco commented:
Thanks for the comments. I'll work on this.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1401 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1401 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 278 insertions(+), 12 deletions(-))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1425 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1425 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 282 insertions(+), 12 deletions(-))

ferrarimarco commented:
I've added all the missing encoders and decoders. I've also implemented the `pre_update`.

I'm still looking at test failures with TF_LOG=DEBUG.

Thanks!

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1554 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1554 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 387 insertions(+), 12 deletions(-))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1554 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1554 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 387 insertions(+), 12 deletions(-))

ferrarimarco commented:
@danawillow I've fixed all your comments, thanks for the review.

Still, I wasn't able to understand why my test fails, although the Cloud Build now passes.

Can you have a look now?

Thanks!

ferrarimarco commented:
PS: I'm running the "basic" test as follows:

```shell
make testacc TF_LOG=DEBUG TEST=./google TESTARGS='-run=TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample'
```

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1552 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1552 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 391 insertions(+), 12 deletions(-))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1552 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1552 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 391 insertions(+), 12 deletions(-))

danawillow commented:
Taking a look through the logs, all the requests/responses seem reasonable, but I did notice this line:
> state_notification_config: was null, but now cty.MapVal(map[string]cty.Value{"pubsub_topic_name":cty.StringVal("")})

Usually we ignore those warnings because they're for things that are basically equivalent, but in this case there's an actual difference: a map with one entry is different than an empty map or nil.

If you take a look at the existing  resource, you can see the following section in the read method:
```
	pubsubTopicName := res.StateNotificationConfig.PubsubTopicName
	if pubsubTopicName != "" {
		d.Set("state_notification_config",
			map[string]string{"pubsub_topic_name": pubsubTopicName})
	} else {
		d.Set("state_notification_config", nil)
	}
```

so sounds like we do typically set the whole map to nil in state if its contents are empty.

ferrarimarco commented:
Nice catch, will check ASAP

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1555 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1555 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 394 insertions(+), 12 deletions(-))

ferrarimarco commented:
Ok, that fixed the test! there are still a couple of failures I'd like to check, but shouldn't be too hard I hope!

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1556 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1556 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 394 insertions(+), 12 deletions(-))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1559 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1559 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 395 insertions(+), 12 deletions(-))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1569 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1569 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 405 insertions(+), 12 deletions(-))

ferrarimarco commented:
@danawillow I've fixed all the issues! Now all the tests pass

There's only one comment that I wasn't able to accomodate: adding `send_empty_value: true` to `logLevel`. If I do that, the `TestAccCloudIoTRegistry_update` test breaks (see my reply in the thread above).

Test output (not verbose):

```shell
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample
=== PAUSE TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample
=== CONT  TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample
--- PASS: TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryBasicExample (7.87s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	9.238s
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryFullExample -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryFullExample
=== PAUSE TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryFullExample
=== CONT  TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryFullExample
--- PASS: TestAccCloudIotDeviceRegistry_cloudiotDeviceRegistryFullExample (15.55s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	16.326s
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccCloudIoTRegistry_update -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccCloudIoTRegistry_update
=== PAUSE TestAccCloudIoTRegistry_update
=== CONT  TestAccCloudIoTRegistry_update
--- PASS: TestAccCloudIoTRegistry_update (23.37s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	24.125s
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestValidateCloudIoTDeviceRegistryId -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestValidateCloudIoTDeviceRegistryId
--- PASS: TestValidateCloudIoTDeviceRegistryId (0.01s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	0.757s
```

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1568 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1568 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 405 insertions(+), 12 deletions(-))

danawillow commented on products/cloudiot/terraform.yaml:
```diff
@@ -0,0 +1,120 @@
+          cloudiot_registry_name: "cloudiot-registry"
+          cloudiot_device_status_topic_name: "default-devicestatus"
+          cloudiot_device_telemetry_topic_name: "default-telemetry"
+          cloudiot_additional_device_telemetry_topic_name: "additional-telemetr"
```
```suggestion
          cloudiot_additional_device_telemetry_topic_name: "additional-telemetry"
```

ferrarimarco replied:
Fixed. This didn't cause any failure because it's just a name :)

good catch!

danawillow commented on products/cloudiot/api.yaml:
```diff
@@ -0,0 +1,194 @@
+          errors will be logged. LogLevel is inclusive; enabling INFO logging
+          will also enable ERROR logging.
+
+          The field allows `NONE`, `ERROR`, `INFO`, `DEBUG`.
```
We recently started documenting enum values automatically, so this can be removed (here and elsewhere)
You can see what it looks like now at https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357#diff-f093bf8c1e1d785bca277d59b605db0bR155-R168

ferrarimarco replied:
Oh nice!

Fixed!.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1565 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1565 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 405 insertions(+), 12 deletions(-))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1567 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1567 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 405 insertions(+), 12 deletions(-))

danawillow commented:
Another thing I just realized- the generated docs only have the generated properties. You'll also need to add custom docs for the handwritten ones.

ferrarimarco commented on products/cloudiot/terraform.yaml:
```diff
@@ -0,0 +1,120 @@
+    legacy_name: "google_cloudiot_registry"
+    import_format: ["{{project}}/locations/{{region}}/registries/{{name}}"]
+    id_format: "projects/{{project}}/locations/{{region}}/registries/{{name}}"
+    self_link: 'projects/{{project}}/locations/{{region}}/registries/{{name}}'
```
Moved in api.yaml

ferrarimarco commented on templates/terraform/examples/cloudiot_device_registry_full.tf.erb:
```diff
@@ -0,0 +1,52 @@
+resource "google_cloudiot_registry" "<%= ctx[:primary_resource_id] %>" {
+  name     = "<%= ctx[:vars]['cloudiot_registry_name'] %>"
+
+  depends_on = [
```
Removed

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1541 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1541 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 405 insertions(+), 12 deletions(-))

ferrarimarco commented:
@danawillow 

> Another thing I just realized- the generated docs only have the generated properties. You'll also need to add custom docs for the handwritten ones.

Added, in terraform.yaml, and verified they're included in the generated document.

Thanks!

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1575 insertions(+), 874 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3357-old..auto-pr-3357) ( 13 files changed, 1575 insertions(+), 874 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3357-old..auto-pr-3357) ( 3 files changed, 405 insertions(+), 12 deletions(-))

ferrarimarco commented:
Nice! Thanks a lot @danawillow !
